### ISSUE TITLE
Do we use LESS or SASS?

### ISSUE THREAD

**Issue by sleicester** [the-open-university/ouice#6]
_From @guyweb on November 14, 2014 13:37_

Is SASS still the preference? http://css-tricks.com/sass-vs-less/

_Copied from original issue: the-open-university/student-theme-framework#6_

**Comment by sleicester** [comment]
_From @stephenturvey on November 14, 2014 18:35_

Neither! Preprocessors try turning CSS into a logic based programming language which is not what it is. Nice, clean, pure CSS gets my vote :)

**Comment by sleicester** [comment]
I agree that we can just jump on the bandwagon on these things, but many say it brings in quite a lot of good features that should be in css.

This seems pretty useful:

```
.common-stuff {
    border: 1px solid red;
    font: ...
}

div.something {
    .common-stuff
}
```
Although I could see some things getting abused rather than thinking about inheritance etc.

**Comment by sleicester** [comment]
_From @stephenturvey on November 14, 2014 21:55_

Yeah, but then you have to scroll up to remember what .common-stuff includes... then you've lost where you were in the code! And then if some other cat wants to edit the file who is familiar with a different preprocessor or isn't familiar with any, you have a problem.

**Comment by sleicester** [comment]
What makes good maintainable CSS?

**Comment by sleicester** [comment]
_From @bgurney on November 17, 2014 16:4_

I can see merit in an atomic methodology using pure CSS (http://patternlab.io/about.html). I do like the idea of each atom/molecule/template having it's own 'namespace/prefix'. It also aligns with the style-guide and the design workflow. I think one problem we have faced in the past is simply locating the right place to put new CSS. Often the files are very long and people end up just adding it to the bottom of the file, which then leads to an unmanageable stylesheet. So this type of pattern makes sense to me if we're trying to solve existing problems. With an atomic methodology I don't think the specific syntax of the style language matters too much and is then probably more about personal preference. Perhaps we need to agree on a methodology rather than controlling the specific 'language'?

**Comment by sleicester** [comment]
_From @bgurney on November 17, 2014 16:39_

Can I also suggest the following conventions, regardless:

1) we use 'inline' media queries for print, mobile etc. rather than adding them onto the link tag (can still have separate files, just wrap the entire file in the media query/ies).

2) we target IE specific CSS through the use of a selector prefix; {html.ie}, {html.ie6}, {html.ie7}, {html.ltie8}, {html.ie8}, {html.ltie9} (where lt = less than) rather than through a separate stylesheet wrapped in conditional tags. As standard our layouts will wrap the HTML tag in conditions instead:

&#x3C;!--[if lt IE 7]&#x3E;&#x3C;html class=&#x22;ie ie6 ltie7 ltie8 ltie9&#x22; lang=&#x22;en&#x22;&#x3E;&#x3C; ![endif]--&#x3E;
&#x3C;!--[if IE 7]&#x3E;&#x3C;html class=&#x22;ie ie7 ltie8 ltie9&#x22; lang=&#x22;en&#x22;&#x3E;&#x3C; ![endif]--&#x3E;
&#x3C;!--[if IE 8]&#x3E;&#x3C;html class=&#x22;ie ie8 ltie9&#x22; lang=&#x22;en&#x22;&#x3E;&#x3C; ![endif]--&#x3E;
&#x3C;!--[if gt IE 8]&#x3E;&#x3C;!--&#x3E;&#x3C;html lang=&#x22;en&#x22;&#x3E;&#x3C; !--&#x3C;![endif]--&#x3E;

Doing this will allow us to:
a) use an Asset Pipeline to serve all of our Javascript and CSS in single files on the Live server.
b) give us better control when targeting quick fixes in our 'custom css' file (hopefully one-day this will be removed if we can get a proper continuous deployment system in place).

**Comment by sleicester** [comment]
_From @stephenturvey on November 17, 2014 17:6_

"Often the files are very long and people end up just adding it to the bottom of the file"

Too true!!

**Comment by sleicester** [comment]
_From @guyweb on November 18, 2014 13:49_

How about you propose a CSS file structure Stephen.

If we create a number of CSS files to avoid one massive one for development purposes, this could be compiled into a single minified one for deployment?

I was thinking something like:

html.css (for just the core HTML elements with no added classes)
layout.css
navigation.css
tables.css
forms.css
images-and-objects.css
boxes.css
lists.css

You'd have all the styles including print and mobile for each component in the files themselves.

Would something like this work?

**Comment by sleicester** [comment]
_From @guyweb on November 18, 2014 13:50_

would probably also need one for headers-and-footers.css

**Comment by sleicester** [comment]
_From @jchapple on November 19, 2014 8:32_

My opinion is that modern CSS is no longer clean and pure, hence the need to use pre-processors at least to some degree (e.g. at least for convenience / shortcuts, not necessarily 'logic'). 

I like the atomic design suggestions. Your file structure looks like a good start, Guy.

**Comment by sleicester** [comment]
It would be good to think of some consistent terminology of the different building blocks of a page. I think the concept of the different granularity and cascading provided by [Atomic Design](http://patternlab.io/about.html) mythology is great, but are 'atoms' and 'molecules' really the terminology we're going to use?
- elements
- components-
- widgets
Are any of these any better?

It's worth getting some separation of the files started as it helps other collaborators get a sense of where to find things, and gives a better sense of change within the Git history.

To do some cross linking, resolving the name of the various granular elements as that will inform the issue on Folder structure #3? And also help give some direction to an Alternative solution for the example site #11.

**Comment by sleicester** [comment]
Following on from some brief chats with @peterdevinegmd this morning, I think there are some situations when CSS becomes intense with calculations and less (or similar) really helps with this as you can express a value as an expression. To get some examples of this would be great.

Sometimes I think visually seeing the nested structure within a .less file does make it easier to see hierarchy, but depending on how you group things normally this is possibly not that much of a win.

**Comment by stuartcrouch** [comment]
Existing tech in IT use LESS. Moodle, ou headers and the digital framework all use Less. Not suggesting an opinion either way, just pointing out that there is knowledge of LESS that could be called upon for support.

**Comment by stuartcrouch** [comment]
I will say I am pro compiler. The agility it gains in rippling out changes out weighs the complexity of learning the preprocessor.

**Comment by sleicester** [comment]
We'd lean towards less for this reason, but think there is a good point by Stephen in that there is a lot you can do with CSS without making things too complicated.

Could do with help in the future demonstrating real benefits against this code base.
